### PR TITLE
fix(router): reject new namespaces / deny namespace auto-creation

### PR DESCRIPTION
**Pull request by domodwyer**
When namespace auto-creation was made optional at runtime (https://github.com/influxdata/influxdb_iox/pull/6346), the namespace autocreator was changed to return an error when the schema cache did not have any entry for the requested namespace.

This unfortunately doesn't work when creating a new namespace via the `NamespaceService`'s `CreateNamespace` gRPC endpoint - the endpoint adds the necessary record to the catalog, but doesn't populate the schema cache once the namespace is created, so subsequent write requests fail to find the namespace in the cache, and reject the write.

Additionally this fails in the presence of multiple routers - if router A creates the namespace, router B won't have it cached it. As is, we'd have to reboot the routers every time a new namespace was created to make them usable! Doh.

**This fix is for both kafkaless, and kafka-based clusters.**

---

* refactor(test): dedicated e2e TestContext module (3efc42baa)
      
      Moves the router's TestContext to its own file/module.

* test(router): initialise gRPC delegate in e2e (6f1869f9d)
      
      Initialise the "rpc mode" gRPC handlers in the router e2e TestContext.

* refactor(test): TestContext::write_lp() helper (ac8fa293c)
      
      Adds a helper method to construct the HTTP write request.

* refactor: expose underlying gRPC implementations (1a7679bce)
      
      Changes the gRPC delegate to return the underlying service (type erased)
      implementations instead of the RPC service wrappers.

* fix: bind NamespaceService to gRPC server (3a9b5a4d2)
      
      I forgot to bind the service!

* refactor: clean up namespace errors (105e35429)
      
      The namespace error was poorly refactored and duplicated the prefix
      string. The "rejected" case is now also tested.

* fix: explicit namespace creation w/ existing ns (7eaa8f59b)
      
      Prior to this commit, namespaces that had been created on one router
      could not be used on another router until the latter was restarted.
      Effectively, newly created namespaces couldn't be used.
      
      After this commit, the catalog is also checked when a cache miss occurs,
      ensuring the router discovers new, not-yet-cached namespaces.

* test(e2e): explicit namespace creation (0aa5469ac)
      
      Adds an end-to-end test of the router's gRPC NamespaceService covering
      creation and reading of new namespaces.

### CODE REVIEW

File: router/src/namespace_resolver/ns_autocreation.rs [hunk 1]
```diff
@@ -91,15 +91,27 @@ where
...
+                    match self.inner.get_namespace_id(namespace).await {
+                        Ok(v) => return Ok(v),
+                        Err(super::Error::Lookup(
+                            iox_catalog::interface::Error::NamespaceNotFoundByName { .. },
+                        )) => {
+                            warn!(%namespace, "namespace not in catalog and auto-creation disabled");
+                            return Err(NamespaceCreationError::Reject(namespace.into()).into());
+                        }
+                        Err(e) => return Err(e),
+                    }
```
Comment (NGA-TRAN) [inline router/src/namespace_resolver/ns_autocreation.rs:108]: 👍

**Review by NGA-TRAN — APPROVED** [review]
🎉 
Yeah, this is a known issued that was raised during retention implementation but was not top priority to get attention at that time due to router-related ingester rewrite. 

I did not look into how this was implemented in router which is not my area of expertise but I did verify it goes and find it when namespace  is not in cache

**Comment by domodwyer** [comment]
> Yeah, this is a known issued

Is there an issue I can close with this too then?

**Comment by cannonpalms** [comment]
Is there a reason that this **MUST** be merged? Migration is made more difficult by it. I would prefer if this remained a facility that _CAN_ be used, but isn't used.

**Comment by domodwyer** [comment]
I don't understand what you're asking - you **want** to have to reboot routers after creating a namespace?! I think maybe you've missed the reasoning here... Slack me!

**Comment by NGA-TRAN** [comment]
> > Yeah, this is a known issued
> 
> > Is there an issue I can close with this too then?
https://github.com/influxdata/influxdb_iox/issues/6175

**Comment by domodwyer** [comment]
Ah that's something else I'm afraid :+1:

**Comment by NGA-TRAN** [comment]
Yeah, but that is the idea

**Pull request merged**